### PR TITLE
feat: getAllUsersByDomain, getFieldIndex

### PR DESCRIPTION
**Pull request by Shivansh-yadav13**
I will add more if this is approved
Signed-off-by: Shivansh-yadav13 <yadavshivansh@gmail.com>

### CODE REVIEW

**Comment by Shivansh-yadav13** [comment]
@nodece @Zxilly can you review this, it is part of rbac with domain

File: src/model/model.ts [hunk 1]
```diff
@@ -83,6 +83,7 @@ export class Model {
     const ast = new Assertion();
     ast.key = key;
     ast.value = value;
+    ast.fieldIndexMap = new Map();
```
Comment (nodece) [inline src/model/model.ts]: Looks like you want to add a cache, I don't suggest you do that.

File: src/model/model.ts [hunk 2]
```diff
@@ -473,4 +474,27 @@ export class Model {
...
   }
+
+  public getFieldIndex(ptype: string, field: string): number {
+    const assertion = this.model.get('p')?.get(ptype);
+    let index = assertion?.fieldIndexMap.get(field);
+    if (index) {
+      return index;
+    }
+    const pattern = `${ptype}_${field}`;
+    index = -1;
```
Comment (nodece) [inline src/model/model.ts]: A easier way:
```js
index = assertion?.tokens.findIndex((n) => n === pattern);
if (index === undefined || index === -1) {
      return -1;
}
assertion?.fieldIndexMap.set(field, index);
return index;
```

File: src/enforcer.ts [hunk 3]
```diff
@@ -189,6 +189,46 @@ export class Enforcer extends ManagementEnforcer {
...
+   *
+   * @param domain the domain.
+   * @returns Array of all users in the domain.
+   */
+  public async getAllUsersByDomain(domain: string): Promise<string[]> {
+    const m = new Map();
+    const g = this.model.model.get('g')?.get('g');
+    const p = this.model.model.get('p')?.get('p');
+    const users: string[] = [];
+    const index = await this.getFieldIndex('p', 'dom');
```
Comment (nodece) [inline src/enforcer.ts:199]: Move to line 199.

**Review by nodece — CHANGES_REQUESTED** [review]

**Comment by hsluoyz** [comment]
@Shivansh-yadav13

File: src/enforcer.ts (hunk 3 above)
Comment (Shivansh-yadav13) [reply src/enforcer.ts:199]: sir, which line to move to 199?

Comment (hsluoyz) [reply src/enforcer.ts:199]: @nodece

Comment (nodece) [reply src/enforcer.ts:199]: 203 and 204 line.

**Comment by hsluoyz** [comment]
@Shivansh-yadav13 plz fix CI errors:

image

File: src/enforcer.ts (hunk 3 above)
Comment (Shivansh-yadav13) [reply src/enforcer.ts:199]: Fixed.

File: src/model/model.ts [hunk 4]
```diff
@@ -473,4 +474,16 @@ export class Model {
     }
     return effected;
   }
+
+  public getFieldIndex(ptype: string, field: string): number {
+    const assertion = this.model.get('p')?.get(ptype);
+    let index = assertion?.fieldIndexMap?.get(field);
```
Comment (nodece) [inline src/model/model.ts]:
```suggestion
  
```

File: src/model/model.ts [hunk 5]
```diff
@@ -473,4 +474,16 @@ export class Model {
...
+    const pattern = `${ptype}_${field}`;
+    index = assertion?.tokens.findIndex((n) => n === pattern);
```
Comment (nodece) [inline src/model/model.ts]:
```suggestion
    const index = assertion?.tokens.findIndex((n) => n === pattern);
```

File: src/model/model.ts [hunk 6]
```diff
@@ -473,4 +474,16 @@ export class Model {
...
+    if (index === undefined || index === -1) {
+      return -1;
+    }
+    assertion?.fieldIndexMap?.set(field, index);
```
Comment (nodece) [inline src/model/model.ts]:
```suggestion

```

File: src/model/model.ts [hunk 7]
```diff
@@ -83,6 +83,7 @@ export class Model {
     const ast = new Assertion();
     ast.key = key;
     ast.value = value;
+    ast.fieldIndexMap;
```
Comment (nodece) [inline src/model/model.ts]:
```suggestion

```

File: src/model/assertion.ts [hunk 8]
```diff
@@ -24,6 +24,7 @@ export class Assertion {
   public tokens: string[];
   public policy: string[][];
   public rm: rbac.RoleManager;
+  public fieldIndexMap: Map<string, number>;
```
Comment (nodece) [inline src/model/assertion.ts]:
```suggestion

```

**Comment by nodece** [comment]
@Shivansh-yadav13 If want to enable the cache, you need to consider removing the cache when the policy has changed.

**Pull request merged**